contracts/cw721-base/src/execute.rs

```diff
@@ -1,38 +1,40 @@
+use cosmwasm_std::{Binary, CustomMsg, Deps, DepsMut, Env, MessageInfo, Response, StdResult};
+use cw2::set_contract_version;
+use cw721::{ContractInfoResponse, Cw721Execute, Cw721ReceiveMsg, Expiration};
 use serde::de::DeserializeOwned;
 use serde::Serialize;
 
-use cosmwasm_std::{Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult};
-
-use cw2::set_contract_version;
-use cw721::{ContractInfoResponse, CustomMsg, Cw721Execute, Cw721ReceiveMsg, Expiration};
-
 use crate::error::ContractError;
-use crate::msg::{ExecuteMsg, InstantiateMsg, MintMsg};
+use crate::msg::{ExecuteMsg, InstantiateMsg, MigrateMsg, MintMsg};
 use crate::state::{Approval, Cw721Contract, TokenInfo};
 
 // Version info for migration
 const CONTRACT_NAME: &str = "crates.io:cw721-base";
 const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 
-impl<'a, T, C, E, Q> Cw721Contract<'a, T, C, E, Q>
+impl<'a, MintExt, ResponseExt, InstantiateExt, ExecuteExt, QueryExt>
+    Cw721Contract<'a, MintExt, ResponseExt, InstantiateExt, ExecuteExt, QueryExt>
 where
-    T: Serialize + DeserializeOwned + Clone,
-    C: CustomMsg,
-    E: CustomMsg,
-    Q: CustomMsg,
+    MintExt: Serialize + DeserializeOwned + Clone,
+    ResponseExt: CustomMsg,
+    InstantiateExt: CustomMsg + DeserializeOwned,
+    ExecuteExt: CustomMsg,
+    QueryExt: CustomMsg,
 {
     pub fn instantiate(
         &self,
         deps: DepsMut,
         _env: Env,
         _info: MessageInfo,
-        msg: InstantiateMsg,
-    ) -> StdResult<Response<C>> {
+        msg: InstantiateMsg<InstantiateExt>,
```

Comment:
```suggestion
        msg: InstantiateMsg<MetadataExt>,
```
`MetadataExt` makes more sense to me here since the generic type is specifically for the metadata.

Reply:
Agreed. Maybe `CollectionMetadataExt` to make it super explicit?